### ISSUE TITLE
Do not allow users to specify names of resources that don't exist yet

### ISSUE THREAD

**Issue by mgibson1121** [pivotal/kpack#255]
This idea stems from accepting  our initial pass at improving validation in #11.  I understand this may be a contentious idea given that it may be a K8s pattern to reference a resource that does not exist because it hasn't been reconciled yet.  But in my view, this is an edge case and it is generally good UX not to allow a user of any product to complicate their experience by referencing entities before they are created.  However, this is just an idea.   

`ServiceAccount` resource:

* Validate that the names of the git or registry secrets specified are known to kpack
    * I generally think we shouldn't let users reference the names of resources that have not yet been declared to avoid potential confusion down the road

`Image` resource:

* Validate that names of the service account and builder referenced are known to kpack
    * Same reason as above

**Comment by matthewmcnew** [comment]
`ServiceAccount` is not a kpack resource so, I don't think we would ever attempt to alter the default kubernetes validation behavior.

**Comment by matthewmcnew** [comment]
Validating that an Image's builder exists is intriguing to me.

**Comment by scothis** [comment]
While I understand and appreciate the desire, this behavior will make installing and using kpack more fought. [Level based triggering](https://hackernoon.com/level-triggering-and-reconciliation-in-kubernetes-1f17fe30333d) is fundamental to Kubernetes controllers and it's worth taking the time to really understand.

There can be many reasons why one resource may not exist at a particular moment. Consider an install where one resource in the install references another resource in the same install. Now there is strict ordering for each resource to be created. Let's suppose the resources are defined in the correct order, not all tools apply resources in the same order they are defined. Let's suppose the resources are applied in the correct order, the API Server may not have fully distributed the state for that resource by the time the dependent resource is created. While replication is fast, it is not instantaneous. Now there is a borked half install that isn't functional.

The correct behavior is to only validate resources based on their structure, and not their content. If when reconciling the resource if a dependent resource doesn't exist, or is not ready, that's ok. You indicate in your status that you are not ready. This is the same as if the resource you were depending on is deleted. When the dependency becomes ready, then the dependent resources can update its own state accordingly.

**Comment by matthewmcnew** [comment]
Agreed, that maintaining "out of order" independence is necessary. Many workflows like installation would become cumbersome to impossible if this was not the case. Along the same lines kpack's reconciliation should not behave differently [ when dependent objects don't exist.](https://github.com/projectriff/system/pull/106)  

However, I believe that this issue is slightly more nuanced. There may be valid usecases in kubernetes that validate the existence of dependent resources such as a namespace in `metadata.namespace` or even the service account in a pod spec. 

I would be intrigued about possibly providing some optional mechanism to validate the existence of some dependent resources for kpack deployments where that use case makes sense.
